common/src/Renderer/EntityDecalRenderer.cpp

```diff
@@ -0,0 +1,484 @@
+{
+  // invalidate any entities that are tracking this brush
+  for (auto& [ent, data] : m_entities)
+  {
+    // skip entities that are going to be recomputed anyway
+    if (!data.validatedBrushes)
+    {
+      continue;
+    }
+
+    // if this brush is tracked by this entity, remove it and recalculate
+    if (data.brushes.count(brushNode))
+    {
+      data.brushes.erase(brushNode);
+      invalidateDecalGeometry(data);
+    }
+  }
+}
+
+void EntityDecalRenderer::invalidateDecalBrushes(EntityDecalData& data) const
+{
+  // do nothing if the brush data is already marked as invalidated
+  if (!data.validatedBrushes)
+  {
+    return;
+  }
+
+  data.validatedBrushes = false;
+
+  // remove the geometry from the VBO as it will be recalculated
+  invalidateDecalGeometry(data);
+}
+
+void EntityDecalRenderer::invalidateDecalGeometry(EntityDecalData& data) const
+{
+  // do nothing if the geometry is already invalid, it's not in the VBO
+  if (!data.validatedGeometry)
+  {
+    return;
+  }
+
+  // if the texture doesn't exist, do nothing
+  // also do nothing if the VBO storage fields are null, but it shouldn't happen
+  if (!data.texture || data.vertexHolderKey == nullptr || data.faceIndicesKey == nullptr)
+  {
+    data.validatedGeometry = false;
+    return;
+  }
+
+  // update the VBO
+  m_vertexArray->deleteVerticesWithKey(data.vertexHolderKey);
+
+  const auto faceIndexHolder = m_faces->at(data.texture);
+  faceIndexHolder->zeroElementsWithKey(data.faceIndicesKey);
+
+  if (!faceIndexHolder->hasValidIndices())
+  {
+    // there are no indices left to render for this texture
+    m_faces->erase(data.texture);
+  }
+
+  data.vertexHolderKey = nullptr;
+  data.faceIndicesKey = nullptr;
+  data.validatedGeometry = false;
+}
+
+void EntityDecalRenderer::validateDecalBrushes(
+  const Model::EntityNode* entityNode, EntityDecalData& data) const
+{
+  if (data.validatedBrushes)
+  {
+    // already validated, no need to check again
+    return;
+  }
+
+  const auto document = kdl::mem_lock(m_document);
+  const auto world = document->world();
+
+  // create a fake brush to represent the entity for intersection purposes
+  // todo: (?) i only do this because collectTouchingNodes only accepts a brush
+  // and i can't find a bbox version of this method
+  const auto bounds = entityNode->physicalBounds();
+  const auto builder = Model::BrushBuilder(
+    world->mapFormat(), document->worldBounds(), document->game()->defaultFaceAttribs());
+  auto brush = builder.createCuboid(bounds, "").value();
+  auto brushNode = Model::BrushNode(brush);
+
+  // collect all the brush nodes that touch the fake brush
+  const auto touching = Model::collectTouchingNodes(
+    std::vector<Model::Node*>{world}, std::vector{&brushNode});
+
+  // track them in the entity
+  data.brushes.clear();
+  for (auto node : touching)
+  {
+    if (const auto* br = dynamic_cast<Model::BrushNode*>(node))
+    {
+      // ignore hidden brushes
+      if (br->visible())
```

Comment:
Btw. I don't think that this is enough to decide whether a node is visible. I think you need to use `EditorContext::visible` for this.